[X86][Clang] VectorExprEvaluator::VisitCallExpr / InterpretBuiltin - add AVX512 KTEST/KORTEST intrinsics to be used in constexpr

rturrado opened the pull request:
Add AVX512 KTEST/KORTEST intrinsics to be used in constexpr.

Fixes #162051

rturrado committed:
Mark instructions as _CONSTEXPR

rturrado committed:
Mark builtin instructions as ConstExpr

rturrado committed:
Update InterpretBuiltin function

rturrado committed:
Add constexpr tests

rturrado committed:
Run git-clang-format

rturrado commented:
I have left a couple of TODOs on purpose. They are more a question to reviewers. I didn't know:
1) If I had to check for a carry/zero flag as a result of a kortestc/kortestz operation, respectively.
2) And if so, how to do it.

rturrado commented:
I have noticed that some builds have failed, so I wonder if I am missing something in the way I'm building:

```
cmake -S llvm -B build -G Ninja -DCMAKE_BUILD_TYPE=Release
ninja -C build check-llvm
```

rturrado committed:
Fix build errors

rturrado committed:
Address comments from RKSimon review

rturrado committed:
Run git-clang-format

RKSimon commented:
You just need `LLVM_ENABLE_PROJECTS='clang'`

rturrado committed:
Trying to fix Linux and Windows builds

Rewrite test_kortest_mask16_u8.
Replace std::pair with kortest/ktest_result structs.

rturrado committed:
Mimic bmi2-builtins.c implementation

Rewrite test_kortest_mask16_u8.
Remove avx512-builtins.h.
Remove kortest/ktest_result structs.

rturrado commented:
A couple of questions.

1) `clang/lib/Headers/avx512fintrin.h` implements both `_mm512_kortestc(A, B)` and `_kortestc_mask16_u8(A, B)` in terms of `__builtin_ia32_kortestchi`. However, `_mm512_kortestc` returns an `int` with the value of `A | B`, while `_kortestc_mask16_u8` returns an `unsigned char` with the value `A | B == 0xFFFF`. Wouldn't it be better to have `mm512_kortestc` also as a builtin?

2) I have updated `clang/lib/AST/ByteCode/InterpBuiltin.cpp` adding cases for all the `ktest` and `kortest` instructions, and calling `interp__builtin_elementwise_int_binop`, which returns an `APInt`. Wouldn't it be better to call `interp__builtin_ia32_test_op` and return a `bool` (which could later be casted to an `unsigned char`)? A possible `_mm512_kortestc` builtin could indeed call `interp__builtin_elementwise_int_binop` and return an `APInt` of 16 bits.

RKSimon reviewed (changes requested):
Still missing handling in ExprConstant.cpp VisitBuiltinCallExpr - we currently have 2 expression interpreters to support

rturrado committed:
Update IntExprEvaluator::VisitBuiltinCallExpr at ExprConstant.cpp

Add ktest/kortest cases.
Change !A&B to ~A&B in InterpretBuiltin at InterpBuiltin.cpp.
Add more ktest cases to InterpretBuiltin.

rturrado committed:
Run git-clang-format

RKSimon commented on clang/test/CodeGen/X86/avx512f-builtins.c:
```diff
@@ -8965,6 +8965,10 @@ int test_mm512_kortestc(__m512i __A, __m512i __B, __m512i __C, __m512i __D) {
 
+TEST_CONSTEXPR(_mm512_kortestc(0x0000, 0x0000) == 0x0000);
+TEST_CONSTEXPR(_mm512_kortestc(0x0000, 0x8000) == 0x8000);
+TEST_CONSTEXPR(_mm512_kortestc(0x0123, 0xFEDC) == 0xFFFF);
```
These should be return 0 or 1 - not a bitmask
https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_mm512_kortestc&ig_expand=3934,3937,3932

rturrado committed:
Fix _m512_kortest tests

rturrado committed:
Fix -Wfree-nonheap-object warnings

rturrado committed:
Run git-clang-format

rturrado commented:
> @rturrado have you been able to investigate the CI crashes?

Hm... weird, `ninja -C build check-lang-codegen-x86` is passing 262 tests for me locally.

But also weird, and it may be related to those CI issues, I'm still getting 2 build warnings locally:

```
[35/72] Building CXX object tools/clang/lib/AST/CMakeFiles/obj.clangAST.dir/ByteCode/InterpBuiltin.cpp.o
In file included from /home/rturrado/Projects/llvm-project/llvm/include/llvm/ADT/APSInt.h:18,
                 from /home/rturrado/Projects/llvm-project/llvm/include/llvm/ADT/APFixedPoint.h:19,
                 from /home/rturrado/Projects/llvm-project/clang/include/clang/AST/APValue.h:17,
                 from /home/rturrado/Projects/llvm-project/clang/lib/AST/ByteCode/Integral.h:16,
                 from /home/rturrado/Projects/llvm-project/clang/lib/AST/ByteCode/Boolean.h:12,
                 from /home/rturrado/Projects/llvm-project/clang/lib/AST/ByteCode/InterpBuiltin.cpp:9:
In destructor ‘llvm::APInt::~APInt()’,
    inlined from ‘clang::interp::APSInt clang::interp::Boolean::toAPSInt() const’ at /home/rturrado/Projects/llvm-project/clang/lib/AST/ByteCode/Boolean.h:50:19,
    inlined from ‘bool clang::interp::interp__builtin_select(InterpState&, CodePtr, const clang::CallExpr*)’ at /home/rturrado/Projects/llvm-project/clang/lib/AST/ByteCode/InterpBuiltin.cpp:2799:7:
/home/rturrado/Projects/llvm-project/llvm/include/llvm/ADT/APInt.h:192:18: warning: ‘void operator delete [](void*)’ called on a pointer to an unallocated object ‘1’ [-Wfree-nonheap-object]
  192 |       delete[] U.pVal;
      |                  ^~~~
In destructor ‘llvm::APInt::~APInt()’,
    inlined from ‘clang::interp::APSInt clang::interp::Boolean::toAPSInt() const’ at /home/rturrado/Projects/llvm-project/clang/lib/AST/ByteCode/Boolean.h:50:19,
    inlined from ‘bool clang::interp::interp__builtin_select(InterpState&, CodePtr, const clang::CallExpr*)’ at /home/rturrado/Projects/llvm-project/clang/lib/AST/ByteCode/InterpBuiltin.cpp:2799:7:
/home/rturrado/Projects/llvm-project/llvm/include/llvm/ADT/APInt.h:192:18: warning: ‘void operator delete [](void*)’ called on a pointer to an unallocated object ‘1’ [-Wfree-nonheap-object]
  192 |       delete[] U.pVal;
      |                  ^~~~
```

However, the build errors seem to come from `ExprConstant.cpp`. I'm probably returning an `unsigned char` instead of an `unsigned int` or something on those lines. I'll have a look!

rturrado committed:
Trying to fix the CI crashes

RKSimon reviewed (approved):
LGTM - cheers
